### ISSUE TITLE
uint64 dataset writing incorrect values

### ISSUE THREAD

**Issue by Achilles1515** [h5py/h5py#1547]
I noticed some of my data being changed while attempting to write a uint64 dataset to an H5 file.
Example of trying to write the same dataset in two ways:

```python
import h5py
import numpy as np

first = 9344170403720592284
second = 3671708271
arr = [first, second]

with h5py.File(r"C:\test.h5", 'w') as h5_file:
    name = "test1"
    h5_file.create_dataset(name, data=arr, dtype=np.uint64)
    bad = h5_file[name][0]
    print(bad)

    name = "test2"
    np_arr = np.array(arr, dtype=np.uint64)  # manually create NumPy array
    h5_file.create_dataset(name, data=np_arr)
    good = h5_file[name][0]
    print(good)
```

Output:
```
9344170403720591360
9344170403720592284
```

In "test1", the last byte of that "bad" uint64 value is being set to 0.
For "test2", I am manually constructing the `np.uint64` array and then passing it into the `create_dataset()` method - this works as expected...though I was expecting "test1" to basically be doing the same thing.

Interestingly,  omitting the "second" value from the list results in the correct output.

```python
import h5py
import numpy as np

first = 9344170403720592284
second = 3671708271
arr = [first]    # no second value

with h5py.File(r"C:\test.h5", 'w') as h5_file:
    # ...same as above
```

Output:
```
9344170403720592284
9344170403720592284
```
Is this all expected behavior?

```
h5py    2.10.0
HDF5    1.10.5
Python  3.8.2 (tags/v3.8.2:7b3ab59, Feb 25 2020, 23:03:10) [MSC v.1916 64 bit (AMD64)]
sys.platform    win32
sys.maxsize     9223372036854775807
numpy   1.18.4
```

**Comment by takluyver** [comment]
It looks like the difference comes from converting the data to a numpy array:

```
In [4]: np.array(arr)                                                                                                  
Out[4]: array([9.34417040e+18, 3.67170827e+09])

In [5]: np.array(arr).dtype                                                                                            
Out[5]: dtype('float64')

In [6]: np.array(arr[:1])                                                                                              
Out[6]: array([9344170403720592284], dtype=uint64)
```

It appears that if you convert one value which can fit in int64 along with one value that can only fit in uint64, numpy chooses float64 for the array. This would appear to contradict the [numpy.array() docs](https://numpy.org/devdocs/reference/generated/numpy.array.html#numpy.array), which say it will use "the minimum type required to hold the objects in the sequence", but I don't know exactly what 'minimum' means.

In most cases, h5py uses `numpy.asarray()` first without specifying a dtype, and then any data conversion happens within HDF5. This isn't necessarily set in stone, but if you pass in an array, it does conveniently avoid us making a temporary in-memory copy of it for the conversion. So I'd be interested to know if numpy considers that a bug before we change anything.

**Comment by tacaswell** [comment]
I agree with @takluyver .

Has this been brought up with the numpy folks yet?

**Comment by takluyver** [comment]
I opened a numpy issue to ask about it: https://github.com/numpy/numpy/issues/16287

**Comment by takluyver** [comment]
It seems to be accepted as a bug with numpy, but not an easy one to fix. Numpy guesses a dtype for each element individually - preferring signed over unsigned integers - and then 'promotes' them together. Promoting int64 and uint64 gives float64, the type which can (approximately) represent any number in the other two.

**Comment by Achilles1515** [comment]
Thanks for bringing this up over at NumPy.
What is the benefit of having NumPy guess the dtype versus just passing the dtype through as an argument if one is specified from the h5py method (as is done in my first example)?

**Comment by takluyver** [comment]
I'm not sure there's a massive benefit - we're already passing it through in one case to work around a putative HDF5 bug. But if you pass in a numpy array along with a different dtype, passing it through to `asarray()` means numpy does the conversion, creating a new array in memory to hold the data. Not passing it through means HDF5 does the conversion, which it should be able to do a chunk at a time.

In short, it's just the result of trying to do the best thing in various different scenarios while avoiding spaghetti code as far as possible, e.g. using `asarray()` rather than branching on 'is it already a numpy array?'

**Comment by takluyver** [comment]
Here's where the relevant code starts, if you want to have a look:

https://github.com/h5py/h5py/blob/2660313fcfdb927303d96848014a639b6230abb0/h5py/_hl/dataset.py#L35

**Comment by tacaswell** [comment]
Given hdf5's seperation of "in memory" and "on disk" data types, I have always understood the dtype in `create_dataset` to be the user specifying the "on disk" version and the "in memory" datatype being defined by what ever you pass in which we have deferred as much as possible to numpy (as at the end of the day we need to have a pointer to a c-style array of a numeric type to pass to libhdf5).  

We only avoid making a copy if your data is already in c-contigious arrays, if the user input comes in as a list we have to make a copy anyway I think it makes sense to give numpy the hint.   

Should be conditioned giving the output type to `np.asarray` on if the input supports the array protocol or not?  

See https://github.com/h5py/h5py/issues/819 for why we have that one case of avoiding letting hdf5 do the casting.

**Comment by takluyver** [comment]
I've opened #1574 to fix this.

I enjoy it when the PR number to fix something is an anagram of the issue number. :slightly_smiling_face:
